ENH: Update read_*_postgis functions.

bifbof opened the pull request:
closes #227 
closes #211
The following changes in this pull request.
- Decorator to handle not only connection strings, but also engines/connection objects
- WKT transformation for locations extent
- Updating accessor arguments (adding missing ones and `if_exists` to "fail".
- new argument `sql` instead of fix string.
- handing building GeoDataFrame to `read_*_gpd` functions

bifbof commented:
There are a lot of warnings, apparently pandas has redesigned its SQL module and geopandas has yet to adapt.

hong2223 commented on trackintel/io/postgis.py:
```diff
@@ -66,56 +94,52 @@ def write_positionfixes_postgis(
     Examples
     --------
-    >>> df.as_positionfixes.to_postgis(conn_string, table_name)
+    >>> pfs.as_positionfixes.to_postgis(conn_string, table_name)
```
would be nice to include also examples of calling this function directly, e.g., 
same for other write functions
```suggestion
    >>> pfs.as_positionfixes.to_postgis(conn_string, table_name)
    >>> ti.io.postgis.write_positionfixes_postgis(pfs, conn_string, table_name)
```

hong2223 commented on trackintel/io/postgis.py:
```diff
@@ -210,11 +230,15 @@ def write_staypoints_postgis(staypoints, conn_string, table_name, schema=None, s
     Examples
     --------
-    >>> df.as_staypoints.to_postgis(conn_string, table_name)
+    >>> spts.as_staypoints.to_postgis(conn_string, table_name)
```
```suggestion
    >>> spts.as_staypoints.to_postgis(conn_string, table_name)
    >>> ti.io.postgis.write_staypoints_postgis(spts, conn_string, table_name)
```

hong2223 commented on trackintel/io/postgis.py:
```diff
@@ -287,50 +303,67 @@ def write_locations_postgis(locations, conn_string, table_name, schema=None, sql
     Examples
     --------
-    >>> df.as_locations.to_postgis(conn_string, table_name)
+    >>> locs.as_locations.to_postgis(conn_string, table_name)
```
```suggestion
    >>> locs.as_locations.to_postgis(conn_string, table_name)
    >>> ti.io.postgis.write_locations_postgis(locs, conn_string, table_name)
```

hong2223 commented on tests/io/test_postgis_gpd.py:
```diff
@@ -246,12 +247,14 @@ def test_read(self, example_positionfixes, conn_postgis):
-            pfs_db = ti.io.read_positionfixes_postgis(conn_string, table, geom_col)
+            pfs_db = ti.io.read_positionfixes_postgis(sql, conn_string, geom_col)
             pfs_db = pfs_db.set_index("id")
+            print(pfs_db)
```
Do we need this printing?

hong2223 commented on tests/io/test_postgis_gpd.py:
```diff
@@ -452,3 +485,30 @@ def test_srid(self, example_positionfixes):
+
+        wrapped(conn)
+
+    pass
```
delete

hong2223 commented on tests/io/test_postgis_gpd.py:
```diff
@@ -452,3 +485,30 @@ def test_srid(self, example_positionfixes):
+        @ti.io.postgis._handle_con_string
+        def wrapped(con):
+            assert con is conn
+            return
```
do we need a return here?

hong2223 commented on tests/io/test_postgis_gpd.py:
```diff
@@ -452,3 +485,30 @@ def test_srid(self, example_positionfixes):
+class Test_Handle_Con_String:
+    def test_conn_string(self, conn_postgis):
+        """Test if decorator opens a connection with connection string and closes it."""
+        conn_string, conn = conn_postgis
```
```suggestion
        conn_string, _ = conn_postgis
```

hong2223 commented on tests/io/test_postgis_gpd.py:
```diff
@@ -452,3 +485,30 @@ def test_srid(self, example_positionfixes):
+
+    def test_conn(self, conn_postgis):
+        """Test handeling of connection input"""
+        conn_string, conn = conn_postgis
```
```suggestion
        _ , conn = conn_postgis
```

hong2223 commented on trackintel/io/postgis.py:
```diff
@@ -22,28 +56,23 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
-    return pfs
+    Examples
+    --------
+    >>> pfs = ti.io.read_postifionfixes("SELECT * FROM postionfixes", con, geom_col="geom")
```
```suggestion
    >>> pfs = ti.io.postgis.read_postifionfixes("SELECT * FROM postionfixes", con, geom_col="geom")
```

hong2223 commented on trackintel/io/postgis.py:
```diff
@@ -66,56 +94,52 @@ def write_positionfixes_postgis(
+    **kwargs
+        Further keyword arguments as available in GeoPanda's GeoDataFrame.from_postgis().
+
     Returns
```
Is it possible to add **Examples** in all _read_ functions?

hong2223 reviewed (approved):
Thanks for the change. One comment (the TODO comment) was left out in the first round, maybe you can have a look? All the others are fine. 

I would request @henrymartin1 for a review since he posted the two issues.

bifbof commented:
> Thanks for the change. One comment (the TODO comment) was left out in the first round, maybe you can have a look? All the others are fine.
> 
> I would request @henrymartin1 for a review since he posted the two issues.

So I looked into it, the saving works but the loading does not so far. The read_*_gpd function has no argument to set the index by column but at least the old index exists as additional column. Me may have to add the remaining arguments for the to_postgis functions (see #243) and then also update the read_*_gpd function.

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -139,55 +167,59 @@ def write_triplegs_postgis(
-        triplegs.to_postgis(table_name, conn, if_exists=if_exists, index=True, chunksize=sql_chunksize)
-    finally:
-        conn.close()
+    triplegs.to_postgis(table_name, con, if_exists=if_exists, schema=schema, index=True, chunksize=sql_chunksize)
```
Maybe we can add the missing arguments from the original geopandas [function](https://geopandas.org/docs/reference/api/geopandas.GeoDataFrame.to_postgis.html) as optional input just in case. 
This would concern these two arguments: 
index_label=None, dtype=None

Our default name for the index is the index name (which I think is very good)

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -210,11 +230,15 @@ def write_staypoints_postgis(staypoints, conn_string, table_name, schema=None, s
+    >>> spts.as_staypoints.to_postgis(conn_string, table_name)
     """
 
     # todo: Think about a concept for the indices. At the moment, an index
```
I think yes. 
So the default should be that if we download something that is not in trackintel format, that it is changed into trackintel format (e.g., the case that there is no index in the database).
But if we download/upload/download something that is in trackintel format, that the table is not changed.

It is a bit tricky to verify this without a test

bifbof replied:
So I looked into it, the saving works but the loading does not so far. The read_*_gpd function has no argument to set the index by column but at least the old index exists as additional column. Me may have to add the remaining arguments for the to_postgis functions (see #243) and then also update the read_*_gpd function.

henrymartin1 reviewed (changes requested):
Really great work! 
I would only propose a minor change so that the `dtype` argument can be passed along to the geopandas functions. 

Also I think it would be good to add two tests to ensure that the open todo is solved by this PR. 
An idea for a test would be the following:
- Stability of trackintel conform tables:
   - There is a dataframe that is trackintel conform in the database. If we download it (->x), upload it and download it again (->y) than x == y should be true.
   - It probably would be better to do the comparison in postgresql/postgis after uploading x again but such a comparison seems a bit harder/weird to do in postgresql ([example1](https://dba.stackexchange.com/questions/72641/checking-whether-two-tables-have-identical-content-in-postgresql/72656), [example2](https://dba.stackexchange.com/questions/206992/check-if-two-tables-have-the-same-structure-are-compatible)

bifbof commented:
With the last two commits, I added all the parameters of `to_postgis` from geopandas to all the write*postgis functions and adjusted our order of arguments accordingly. Since all functions use exactly the same function in the same way, I took the liberty to simplify the handling with the docstrings for these functions a bit.

I haven't done the index preservation test yet. For that you would have to change all read*postgis functions and the related read*gpd functions. As this PR is already big, I would move that to another PR.

henrymartin1 commented:
> 
> 
> With the last two commits, I added all the parameters of `to_postgis` from geopandas to all the write*postgis functions and adjusted our order of arguments accordingly. Since all functions use exactly the same function in the same way, I took the liberty to simplify the handling with the docstrings for these functions a bit.
> 
> I haven't done the index preservation test yet. For that you would have to change all read*postgis functions and the related read*gpd functions. As this PR is already big, I would move that to another PR.

Sounds good. Could you please open an issue with a description of the problem for this?

bifbof commented:
> > With the last two commits, I added all the parameters of `to_postgis` from geopandas to all the write*postgis functions and adjusted our order of arguments accordingly. Since all functions use exactly the same function in the same way, I took the liberty to simplify the handling with the docstrings for these functions a bit.
> > I haven't done the index preservation test yet. For that you would have to change all read_postgis functions and the related read_gpd functions. As this PR is already big, I would move that to another PR.
> 
> Sounds good. Could you please open an issue with a description of the problem for this?

Already did. :)

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -22,348 +56,259 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
+    --------
+    >>> pfs = ti.io.read_postifionfixes_postgis("SELECT * FROM postionfixes", con, geom_col="geom")
+    """
+    pfs = gpd.GeoDataFrame.from_postgis(sql, con, geom_col, *args, **kwargs)
```
This is OK

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -22,348 +56,259 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
 ):
-    """Stores positionfixes to PostGIS. Usually, this is directly called on a positionfixes
-    DataFrame (see example below).
+    positionfixes.to_postgis(name, con, schema, if_exists, index, index_label, chunksize, dtype)
```
Here we should use keyword arguments to tunnel these many arguments to prevent future errors.

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -22,348 +56,259 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
-
-def read_triplegs_postgis(conn_string, table_name, geom_col="geom", *args, **kwargs):
+@_handle_con_string
+def read_triplegs_postgis(sql, con, geom_col="geom", *args, **kwargs):
```
It is not allowed to have arguments after an keyword argument

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -22,348 +56,259 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
+    --------
+    >>> tpls = ti.io.read_triplegs_postgis("SELECT * FROM triplegs", con, geom_col="geom")
+    """
+    tpls = gpd.GeoDataFrame.from_postgis(sql, con, geom_col=geom_col, index_col="id", *args, **kwargs)
```
It is not allowed to use positional arguments (*args) after a keyword argument (index_col="id")

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -22,348 +56,259 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
-
-    table_name : str
-        The name of the table to write to.
+    triplegs.to_postgis(name, con, schema, if_exists, index, index_label, chunksize, dtype)
```
I think we should use keyword arguments to pass on so many arguments

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -22,348 +56,259 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
-
-def read_staypoints_postgis(conn_string, table_name, geom_col="geom", *args, **kwargs):
+@_handle_con_string
+def read_staypoints_postgis(sql, con, geom_col="geom", *args, **kwargs):
```
It is not allowed to use positional arguments (*args) after a keyword argument

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -22,348 +56,259 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
+def write_staypoints_postgis(
+    staypoints, name, con, schema=None, if_exists="fail", index=True, index_label=None, chunksize=None, dtype=None
+):
+    staypoints.to_postgis(name, con, schema, if_exists, index, index_label, chunksize, dtype)
```
I would use kwds

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -22,348 +56,259 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
 
-def read_locations_postgis(conn_string, table_name, geom_col="geom", *args, **kwargs):
+@_handle_con_string
+def read_locations_postgis(sql, con, geom_col="geom", *args, **kwargs):
```
It is not allowed to use positional arguments (*args) after a keyword argument

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -22,348 +56,259 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
-        )
-    finally:
-        conn.close()
+    locs = gpd.GeoDataFrame.from_postgis(sql, con, geom_col=geom_col, index_col="id", *args, **kwargs)
```
It is not allowed to use positional arguments (*args) after a keyword argument

henrymartin1 commented on trackintel/io/postgis.py:
```diff
@@ -22,348 +56,259 @@ def read_positionfixes_postgis(conn_string, table_name, geom_col="geom", *args,
+    >>> trips = ti.io.read_trips_postgis("SELECT * FROM trips", con, geom_col="geom")
 
+    """
+    trips = pd.read_sql(sql, con, index_col="id", *args, **kwargs)
```
It is not allowed to use positional arguments (*args) after a keyword argument

henrymartin1 reviewed (changes requested):
Hey nice work, fancy idea with the doc strings ;-) 
I would leave them now like they are but I think in general it is a bit weird to not have docstrings directly in the code. I think a better alternative would be to link the geopandas documentation for further information.

I like the minimalistic style for the reading function to simply pass everything to the geopandas function but I would change it to keyword assignments to exclude positional errors. 
I wanted to fix this myself but then I found a different error where we assign keyword arguments and then pass `*args` which I think is not allowed. I have marked these location in the code

henrymartin1 reviewed (approved):
Great work, thanks for the PR!
